bootstrap/app_test.go

```diff
@@ -0,0 +1,139 @@
+	core "authz/api/gen/v1alpha"
+	"authz/api/grpc"
+	"context"
+	"os"
+	"path"
+	"path/filepath"
+	"runtime"
+	"strconv"
+	"sync/atomic"
+	"testing"
+
+	"github.com/ory/dockertest/v3"
+	"github.com/stretchr/testify/assert"
+	"google.golang.org/grpc/metadata"
+)
+
+// These smoketests should exercise minimal functionality in vertical slices, primarily to ensure correct startup
+
+func TestCheckAccess(t *testing.T) {
+	t.Parallel()
+	srv := initializeGrpcServer(t)
+
+	resp, err := srv.CheckPermission(getContext(), &core.CheckPermissionRequest{
+		Subject:      "u1",
+		Operation:    "access",
+		Resourcetype: "license",
+		Resourceid:   "o1/smarts",
+	})
+
+	assert.NoError(t, err)
+
+	assert.True(t, resp.Result)
+}
+
+func TestGetLicense(t *testing.T) {
+	t.Parallel()
+	srv := initializeGrpcServer(t)
+
+	resp, err := srv.GetLicense(getContext(), &core.GetLicenseRequest{
+		OrgId:     "o1",
+		ServiceId: "smarts",
+	})
+
+	assert.NoError(t, err)
+
+	assert.EqualValues(t, 9, resp.SeatsAvailable)
+	assert.EqualValues(t, 10, resp.SeatsTotal)
+}
+
+func TestGetAssigned(t *testing.T) {
+	t.Parallel()
+	srv := initializeGrpcServer(t)
+
+	includeUsers := true
+	filter := core.SeatFilterType_assigned
+	resp, err := srv.GetSeats(getContext(), &core.GetSeatsRequest{
+		OrgId:        "o1",
+		ServiceId:    "smarts",
+		IncludeUsers: &includeUsers,
+		Filter:       &filter,
+	})
+
+	assert.NoError(t, err)
+
+	assert.EqualValues(t, 1, len(resp.Users))
+}
+
+func TestModify(t *testing.T) {
+	t.Parallel()
+	srv := initializeGrpcServer(t)
+
+	_, err := srv.ModifySeats(getContext(), &core.ModifySeatsRequest{
+		OrgId:     "o1",
+		ServiceId: "smarts",
+		Assign:    []string{"u2"},
+		Unassign:  []string{"u1"},
+	})
+
+	assert.NoError(t, err)
+}
+
+func initializeGrpcServer(t *testing.T) *grpc.Server {
+	token, err := serialKey()
+	assert.NoError(t, err)
+
+	grpc, _ := initialize("localhost:"+port, token, "spicedb", false)
+
+	return grpc
+}
+
+func getContext() context.Context {
+	data := metadata.New(map[string]string{
+		"grpcgateway-authorization": "token",
+	})
+
+	return metadata.NewIncomingContext(context.Background(), data)
+}
+
+var port string
+
+func TestMain(m *testing.M) {
+	pool, err := dockertest.NewPool("") // Empty string uses default docker env
+	if err != nil {
+		return
+	}
+
+	var (
+		_, b, _, _ = runtime.Caller(0)
+		basepath   = filepath.Dir(b)
+	)
+
+	resource, err := pool.RunWithOptions(&dockertest.RunOptions{
+		Repository:   "authzed/spicedb",
+		Tag:          "v1.17.0", // Replace this with an actual version
+		Cmd:          []string{"serve-testing", "--load-configs", "/mnt/spicedb_bootstrap.yaml"},
+		Mounts:       []string{path.Join(basepath, "../schema/spicedb_bootstrap.yaml") + ":/mnt/spicedb_bootstrap.yaml"},
+		ExposedPorts: []string{"50051/tcp", "50052/tcp"},
+	})
+	if err != nil {
+		return
+	}
+
+	port = resource.GetPort("50051/tcp")
+
+	result := m.Run()
+	_ = pool.Purge(resource)
+
+	os.Exit(result)
+}
+
+var keyData int32 = 1
```

Comment:
same comment - probably my personal choice, but if you feel this is better to be declared at the top its fine. <either way i am ok>. But this variable should be a "const" as it is not intended to be changed.

Reply:
It increments for each test and is only used by the function below it. This was chosen instead of the random key because, despite it seeming like a statistic impossibility, I had a test failure that looked like pollution from one test to another, and the only way I can see that could happen is if two tests somehow had the same access key. Randomized keys _should_ never repeat like that, atomically incremented ones _will_ never repeat like that.

Reply:
Oh i missed the increment part, i think its ok for now then.

Reply:
Yeah. It _could_ make sense to make it its own struct or something, but that seems a little extra imo. As-is, it's essentially a static field that gets atomically incremented.

Reply:
Also missed the atomic part, which is take care of concurrency while increments. closing the comment for now